core/node/rpc/sync/client/syncer_set.go

```diff
@@ -670,46 +672,118 @@ func (ss *SyncerSet) rmStream(streamID StreamId) {
-	}
+	var invalid bool
 
-	// TODO: Add backfill validation
+	// Prevent duplicates in the backfill list
+	seen := make(map[StreamId]struct{})
+	for _, backfill := range mr.ToBackfill {
+		for _, c := range backfill.GetStreams() {
+			streamId, err := StreamIdFromBytes(c.GetStreamId())
+			if err != nil {
+				mr.BackfillingFailureHandler(&SyncStreamOpStatus{
+					StreamId: c.GetStreamId(),
+					Code:     int32(Err_INVALID_ARGUMENT),
+					Message:  "Invalid stream ID in backfill operation",
+				})
+				invalid = true
+				continue
+			}
 
-	// Prevent duplicates in the add list
-	if len(mr.ToAdd) > 1 {
-		seen := make(map[StreamId]struct{}, len(mr.ToAdd))
-		for _, c := range mr.ToAdd {
-			streamId := StreamId(c.GetStreamId())
 			if _, exists := seen[streamId]; exists {
-				return RiverError(Err_INVALID_ARGUMENT, "Duplicate stream in add operation")
+				mr.BackfillingFailureHandler(&SyncStreamOpStatus{
+					StreamId: c.GetStreamId(),
+					Code:     int32(Err_INVALID_ARGUMENT),
+					Message:  "Duplicate stream in backfill operation",
+				})
+				invalid = true
+				continue
 			}
 			seen[streamId] = struct{}{}
 		}
 	}
 
+	// Prevent duplicates in the add list
+	seen = make(map[StreamId]struct{}, len(mr.ToAdd))
+	for _, c := range mr.ToAdd {
+		streamId, err := StreamIdFromBytes(c.GetStreamId())
+		if err != nil {
+			mr.AddingFailureHandler(&SyncStreamOpStatus{
+				StreamId: c.GetStreamId(),
+				Code:     int32(Err_INVALID_ARGUMENT),
+				Message:  "Invalid stream ID in adding operation",
+			})
+			invalid = true
+			continue
+		}
+
+		if _, exists := seen[streamId]; exists {
+			mr.AddingFailureHandler(&SyncStreamOpStatus{
+				StreamId: c.GetStreamId(),
+				Code:     int32(Err_INVALID_ARGUMENT),
+				Message:  "Duplicate stream in adding operation",
+			})
+			invalid = true
+			continue
+		}
+		seen[streamId] = struct{}{}
+	}
+
 	// Prevent duplicates in the remove list
-	if len(mr.ToRemove) > 1 {
-		seen := make(map[StreamId]struct{}, len(mr.ToRemove))
-		for _, s := range mr.ToRemove {
-			streamId := StreamId(s)
-			if _, exists := seen[streamId]; exists {
-				return RiverError(Err_INVALID_ARGUMENT, "Duplicate stream in remove operation")
+	seen = make(map[StreamId]struct{}, len(mr.ToRemove))
+	for _, s := range mr.ToRemove {
+		streamId, err := StreamIdFromBytes(s)
+		if err != nil {
+			mr.RemovingFailureHandler(&SyncStreamOpStatus{
+				StreamId: s,
+				Code:     int32(Err_INVALID_ARGUMENT),
+				Message:  "Invalid stream ID in backfill operation",
+			})
+			invalid = true
+			continue
+		}
+
+		if _, exists := seen[streamId]; exists {
+			mr.RemovingFailureHandler(&SyncStreamOpStatus{
+				StreamId: s,
+				Code:     int32(Err_INVALID_ARGUMENT),
+				Message:  "Duplicate stream in removing operation",
+			})
+			invalid = true
+			continue
+		}
+		seen[streamId] = struct{}{}
+	}
+
+	// Prevent passing the same stream to both add and remove operations
+	if !invalid {
```

Comment:
I'm not sure if this is correct? If `invalid=true` the agreement is that the caller stops execution and returns nil. In this case it can miss reports of streams that are both in the add and remove request?